streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -1193,6 +1196,11 @@ public void updateCommittedOffsets(final TopicPartition topicPartition, final Lo
         committedOffsets.put(topicPartition, offset);
     }
 
+    @Override
+    public void updateEndOffsets(final TopicPartition topicPartition, final Long offset) {
+        highWatermark.put(topicPartition, offset);
```

Comment:
Does this actually improve the endOffsets accuracy? This part of the code is new to me so I just had a look around, and I'm a bit confused about how it's supposed to work. Here's my reading of the code: 

the endOffsets are exposed to the user via the TaskMetadata, which is updated/created in StreamThread#updateThreadMetadata. That method seems to be called only when the StreamThread's state is set to `RUNNING`, which in turn only occurs when it completes restoration. So it would seem that it never gets updated during normal processing, when the end offset is actually increasing. In the ticket you say it's not updated "optimally", is it possible that it's not updated at all? 

Also, assuming the task metadata update is refactored to occur more regularly, I'm not sure it helps anything to  update the `highWatermark` more quickly/frequently since this is only read out and used when 
 it's returned by the `highWaterMark()` method, which is where it was being updated before. So it might be that we don't need this part of it at all, and just refactoring the TaskMetadata update alone will be sufficient

Reply:
Oh I see, I thought we were making a copy of the end offsets in StreamTask#highWatermark, but it's just an unmodifiableMap